crates/components/poseidon-halo2/src/spec.rs

```diff
@@ -0,0 +1,108 @@
+use crate::{rate15_params, rate1_params, rate2_params};
+use group::ff::Field;
+use halo2_poseidon::poseidon::primitives::Spec;
+use halo2_proofs::halo2curves::bn256::Fr as F;
+
+/// The type used to hold the MDS matrix and its inverse.
+pub(crate) type Mds<F, const T: usize> = [[F; T]; T];
+
+/// The spec for rate-15 Poseidon used both inside the circuit and in the clear.
+// Patterned after https://github.com/privacy-scaling-explorations/poseidon-gadget/blob/764a682ee448bfbde0cc92a04d241fe738ba2d14/src/poseidon/primitives/p128pow5t3.rs#L15
+#[derive(Debug)]
+pub struct Spec15;
+
+impl Spec<F, 16, 15> for Spec15 {
+    fn full_rounds() -> usize {
+        8
+    }
+
+    fn partial_rounds() -> usize {
+        // Taken from https://github.com/iden3/circomlib/blob/master/circuits/poseidon.circom
+        // (see "var N_ROUNDS_P[16]"), where they use 64 partial rounds for 15-rate Poseidon
+        64
+    }
+
+    fn sbox(val: F) -> F {
+        val.pow_vartime([5])
```

Comment:
nit: 
```
let a = val * val;

let b = a * a;

b * val
```
is much faster than val.pow_vartime([5]) ..ymmv

Reply:
thanks, i'll keep it as `val.pow_vartime([5])` for transparency since that's how `scroll-tech`'s poseidon computes it.